burp-dradis.rb

```diff
@@ -447,48 +436,68 @@ def create_dradis_issue(issue)
     token    = @field_token.text    || ''
     payload  = build_json_payload(issue)
 
-    path = ''
-    path << @field_path.text || '' if @radio_pro.selected
-    path << '/' unless path[-1,1] == '/'
-    path << 'api/issues'
-
     unless endpoint.length > 0 && token.length > 0
       javax.swing.JOptionPane.showMessageDialog(nil, "Please configure the extension using the #{META::TAB_CAPTION} tab.")
       return
     end
 
     begin
-      uri      = URI.parse(endpoint)
-      http     = Net::HTTP.new(uri.host, uri.port)
-      request  = Net::HTTP::Post.new(path)
-
-      request['Content-Type']      = 'application/json'
+      t = build_http_request_thread(endpoint, token, payload)
+      t.start
+    rescue Exception => e
+      @callbacks.issue_alert("There was an error connecting to Dradis: #{e.message}.")
+      @stderr.println e.backtrace
+    end
+  end
 
+  # Internal: builds a Thread object that makes an HTTP request using the Burp API
+  #
+  # endpoint - The configured Dradis endpoint
+  # token - The configured Dradis API token
+  # payload - The HTTP request body to be sent
+  #
+  # Returns a Thread object
+  #
+  def build_http_request_thread endpoint, token, payload
+    port, use_ssl=
       if endpoint =~ /\Ahttps/i
-        http.use_ssl     = true
-        http.verify_mode = OpenSSL::SSL::VERIFY_NONE if @check_ignore_ssl_errors.selected
+        [443, true]
+      else
+        [80, false]
       end
 
+    path = endpoint.clone
+    path << @field_path.text || '' if @radio_pro.selected
+    path << '/' unless path[-1,1] == '/'
+    path << 'api/issues'
+
+    auth_header =
       if @radio_pro.selected
-        request['Authorization']     = "Token token=\"#{token}\""
-        request['Dradis-Project-Id'] = @field_project_id.text
+        "Authorization: Token token=\"#{token}\"\n" \
+        "Dradis-Project-Id: #{@field_project_id.text}"
       else
-        request.basic_auth('BurpExtender', token)
+        basic_token = Base64.getEncoder().encodeToString("BurpExtender:#{token}".to_java_bytes)
+        "Authorization: Basic #{basic_token}"
       end
 
-      request.body = payload
+    return java.lang.Thread.new(Proc.new {
+      host = endpoint.sub(/^https?\:\/\//, '')
+      url = java.net.URL.new(path)
 
+      # Build the initial request
```

Comment:
Using `build_http_message` as opposed to `build_http_request` gives me the ability to do the header adding but the initial http method, url, and host is omitted from the http message. This means I'll have to the ugly concatenation regardless.